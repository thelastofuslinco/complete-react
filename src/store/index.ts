import { configureStore } from '@reduxjs/toolkit'
import counterSlice, {
  increment,
  decrement,
  incrementByAmount
} from './slices/counterSlice'
import { reset } from './actions'
import expensesSlice, {
  addExpense,
  filterExpenses
} from './slices/expensesSlice'

const store = configureStore({
  reducer: {
    counter: counterSlice,
    expenses: expensesSlice
  },
  devTools: true
})

export type RootState = ReturnType<typeof store.getState>

export {
  store,
  increment,
  decrement,
  incrementByAmount,
  reset,
  addExpense,
  filterExpenses
}
