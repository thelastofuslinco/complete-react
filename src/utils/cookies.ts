export function setCookie(name, value, days) {
  const d = new Date()
  d.setTime(d.getTime() + days * 24 * 60 * 60 * 1000)
  const expires = 'expires=' + d.toUTCString()
  document.cookie = `${name}=${value};${expires};path=/;Secure;HttpOnly;SameSite=Strict`
}

export function getCookie(cookieName) {
  const name = cookieName + '='
  const ca = document.cookie.split(';')

  for (let i = 0; i < ca.length; i++) {
    let c = ca[i]
    while (c.charAt(0) == ' ') {
      c = c.substring(1)
    }
    if (c.indexOf(name) == 0) {
      return c.substring(name.length, c.length)
    }
  }
  return ''
}
