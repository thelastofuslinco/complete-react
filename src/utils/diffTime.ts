export const diffTime = (time) => {
  const date1 = new Date(time).getTime()
  const date2 = new Date().getTime()
  const diff = Math.abs(date2 - date1)
  const diffTime = {
    s: Math.floor(diff / 1000),
    m: Math.floor(diff / (1000 * 60)),
    h: Math.floor(diff / (1000 * 60 * 60))
  }

  if (diffTime.h > 1) {
    return `${diffTime.h} hour ago`
  } else if (diffTime.m > 1) {
    return `${diffTime.m} minutes ago`
  } else {
    return `${diffTime.s} seconds ago`
  }
}
