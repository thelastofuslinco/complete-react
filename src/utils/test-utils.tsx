import { render } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import expensesSlice from '../store/slices/expensesSlice'

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { expenses: expensesSlice },
      preloadedState
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }

  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions })
  }
}
