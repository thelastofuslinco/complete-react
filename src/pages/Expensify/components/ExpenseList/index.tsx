import { ConnectedProps, connect } from 'react-redux'
import { RootState } from '../../../../store'
import ExpenseItem from '../ExpenseItem'

const ExpenseList = ({ data, filters }: PropsFromRedux) => {
  const filteredExpenses = data
    .filter((expense) => {
      const createdDate = new Date(expense.createdAt).setHours(0, 0, 0, 0)
      const textMatch = expense.description.includes(filters.text)
      const startDateMatch = filters.startDate
        ? createdDate >= new Date(filters.startDate).setHours(0, 0, 0, 0)
        : true

      const endDateMatch = filters.endDate
        ? createdDate <= new Date(filters.endDate).setHours(0, 0, 0, 0)
        : true

      return textMatch && startDateMatch && endDateMatch
    })
    .sort((a, b) => {
      if (filters.sortBy === 'amount asc') {
        return a.amount - b.amount
      } else if (filters.sortBy === 'amount desc') {
        return b.amount - a.amount
      } else if (filters.sortBy === 'description asc') {
        if (a.description < b.description) return -1
        else if (a.description > b.description) return 1
      } else if (filters.sortBy === 'description desc') {
        if (a.description < b.description) return 1
        else if (a.description > b.description) return -1
      } else if (filters.sortBy === 'createdAt asc') {
        return new Date(a.createdAt).getTime() - new Date(b.createdAt).getTime()
      } else if (filters.sortBy === 'createdAt desc') {
        return new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime()
      }

      return 0
    })

  return (
    <ul className="expenseList">
      {filteredExpenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        filteredExpenses.map((expense) => (
          <ExpenseItem {...expense} key={expense.id} />
        ))
      )}
    </ul>
  )
}

const conector = connect((state: RootState) => state.expenses)

type PropsFromRedux = ConnectedProps<typeof conector>

export default conector(ExpenseList)
