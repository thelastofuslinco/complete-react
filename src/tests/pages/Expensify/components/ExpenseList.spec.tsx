import React from 'react'
import { screen } from '@testing-library/react'
import ExpenseList from '../../../../pages/Expensify/components/ExpenseList'
import { renderWithProviders } from '../../../../utils/test-utils'
import { expenses } from '../../../fixtures/expenses'
import { filterExpenses } from '../../../../store'

describe('Expense list component', () => {
  test('should render expense list', () => {
    const { store } = renderWithProviders(<ExpenseList />, {
      preloadedState: {
        expenses: {
          data: expenses,
          filters: {
            text: 'a',
            sortBy: 'description desc',
            startDate: '11/04/2023',
            endDate: '11/05/2023'
          }
        }
      }
    })
    store.dispatch(
      filterExpenses({
        text: 'a',
        sortBy: 'amount desc',
        startDate: '11/04/2023',
        endDate: '11/05/2023'
      })
    )
    const element = screen.getByRole('list')
    expect(element).toMatchSnapshot()
  })
})
