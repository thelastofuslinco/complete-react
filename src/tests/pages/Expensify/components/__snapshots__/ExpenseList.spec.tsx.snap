// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense list component should render expense list 1`] = `
<ul
  class="expenseList"
>
  <li
    class="expenseContainer"
  >
    <span>
      Description: 
      any_description
    </span>
    <span>
      Amount: 
      822
    </span>
    <span>
      Note: 
      any_description
    </span>
    <span>
      CreatedAt: 
      11/5/2023, 12:00:00 AM
    </span>
    <button>
      Edit
    </button>
  </li>
  <li
    class="expenseContainer"
  >
    <span>
      Description: 
      any_description
    </span>
    <span>
      Amount: 
      98
    </span>
    <span>
      Note: 
      any_description
    </span>
    <span>
      CreatedAt: 
      11/4/2023, 12:00:00 AM
    </span>
    <button>
      Edit
    </button>
  </li>
</ul>
`;
